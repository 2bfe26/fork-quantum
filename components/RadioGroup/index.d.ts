import { Component, ReactNode, ChangeEventHandler, ComponentType } from 'react';
import { IconNames } from '../Icon';

export interface RadioProps {
  disabled?: boolean;
  error?: boolean;
  children?: string | ReactNode[] | ReactNode;
  label?: string;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  theme?: {
    colors?: {};
    spacing?: {};
  };
  value: string;
}

export type Radio = ComponentType<RadioProps>;

export interface RadioButtonProps {
  checked?: boolean;
  children?: string;
  skin?: 'neutral' | 'primary' | 'success' | 'warning' | 'error';
  size?: 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge';
  disabled?: boolean;
  error?: boolean;
  icon?: IconNames | string;
  id?: string;
  inline?: boolean;
  label?: string;
  onChange?: () => void | ChangeEventHandler<HTMLInputElement>;
  theme?: {
    baseFontSize?: number;
    spacing?: {};
    colors?: {};
    components?: {
      button?: {};
    };
  };
  value: string;
}

export type RadioButton = ComponentType<RadioButtonProps>;

export interface RadioGroupProps {
  type?: 'radio' | 'button';
  size?: 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge';
  options?: Array<{
    label?: ReactNode;
    value: string;
    disabled?: boolean;
  }>;
  children?: ReactNode[] | ReactNode;
  inline?: boolean;
  onChange?: () => {} | ChangeEventHandler<HTMLInputElement>;
  defaultValue?: string;
  name: string;
  error?: string;
  theme?: {
    colors?: {};
    spacing?: {};
  };
  required?: boolean;
}

export default class RadioGroup extends Component<RadioGroupProps> {
  static Radio: Radio;

  static Button: RadioButton;
}
