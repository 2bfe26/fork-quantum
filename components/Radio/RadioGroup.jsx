import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Group = styled.div`
  align-items: flex-start;
  display: flex;
  flex-direction: column;

  span {
    margin-bottom: 20px;
  }

  ${props => props.inline && `
    align-items: unset;
    flex-direction: unset;
  `}
`;

/**
 * Group for Radio components.
 */
const RadioGroup = props => <Group {...props} />;

RadioGroup.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
};

export default RadioGroup;
