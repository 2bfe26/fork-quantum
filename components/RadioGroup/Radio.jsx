import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Colors from '../Colors';
import Label from '../shared/Label';
import HiddenInput from '../shared/HiddenInput';

const RadioMark = styled.span`
  background-color: ${Colors.WHITE};
  border: 1.5px solid ${Colors.BLACK[400]};
  border-radius: 50%;
  box-sizing: border-box;
  display: inline-block;
  height: 16px;
  margin-right: 5px;
  position: relative;
  top: 4px;
  width: 16px;

  :after {
    background-color: ${Colors.BLUE[500]};
    border-radius: 50%;
    content: '';
    display: none;
    height: 50%;
    left: 25%;
    position: absolute;
    top: 25%;
    width: 50%;
  }
`;

const RadioLabel = styled(Label)`
  cursor: pointer;
  user-select: none;

  ${HiddenInput} {
    :checked {
      ~ ${RadioMark} {
        border-color: ${Colors.BLUE[500]};
        :after {
          background-color: ${Colors.BLUE[500]};
          display: block;
        }
      }
    }

    :focus {
      ~ ${RadioMark} {
        border-color: ${Colors.BLUE[500]};
        box-shadow: 0 2px 6px 0 ${Colors.BLUE[50]};
      }
    }
  }

  :hover,
  :focus {
    ${RadioMark} {
      border-color: ${Colors.BLUE[500]};
      box-shadow: 0 2px 6px 0 ${Colors.BLUE[50]};
    }
  }

  ${({ error }) =>
    error &&
    `
    ${HiddenInput} {
      :checked {
        ~ ${RadioMark} {
          border-color: ${Colors.ERROR[500]};
          :after {
            background-color: ${Colors.ERROR[500]};
            display: block;
          }
        }
      }
      :focus {
        ~ ${RadioMark} {
          border-color: ${Colors.ERROR[500]};
          box-shadow: 0 2px 6px 0 ${Colors.ERROR[500]};
        }
      }
    }

    ${RadioMark} {
      border-color: ${Colors.ERROR[500]};
    }

    :hover,
    :focus {
      ${RadioMark} {
        border-color: ${Colors.ERROR[500]};
        box-shadow: 0 2px 6px 0 ${Colors.ERROR[500]};
      }
    }
    `}



  ${({ disabled }) =>
    disabled &&
    `
    color: ${Colors.BLACK[400]};

    :hover {
      cursor: not-allowed;
    }

    ${HiddenInput} {
      :disabled {
        ~ ${RadioMark} {
          background-color: ${Colors.BLACK[200]};
        }
      }

      :checked {
        ~ ${RadioMark} {
          border-color: ${Colors.ERROR[500]};
          :after {
            background-color: ${Colors.ERROR[500]};
            display: block;
          }
        }
      }

      :checked:disabled {
        background-color: ${Colors.BLACK[400]};
        border-color: ${Colors.BLACK[400]};
        ~ ${RadioMark}:after {
          background-color: ${Colors.BLACK[400]};
        }
      }

      :focus {
        ~ ${RadioMark} {
          border-color: ${Colors.BLACK[400]};
          box-shadow: none;
        }
      }
    }

    :hover, :focus {
      ${RadioMark} {
        border-color: ${Colors.BLACK[400]};
        box-shadow: none;
      }
    }
  `}

  ${HiddenInput}:checked:disabled ~ ${RadioMark} {
    background-color: ${Colors.BLACK[400]};
    border-color: ${Colors.BLACK[400]};
    box-shadow: inset 0 0 0 3.5px ${Colors.WHITE};
  }
`;

const Radio = ({
  children,
  label,
  error,
  disabled,
  onChange,
  value,
  ...rest
}) => (
  <RadioLabel error={error} disabled={disabled}>
    <HiddenInput
      type="radio"
      disabled={disabled}
      onChange={e => onChange({ value, label }, e)}
      value={value}
      {...rest}
    />
    <RadioMark />
    {children || label}
  </RadioLabel>
);

Radio.displayName = 'RadioGroup.Radio';

Radio.defaultProps = {
  disabled: false,
  error: false,
  children: undefined,
  label: undefined,
  onChange: () => {},
};

Radio.propTypes = {
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  children: PropTypes.string,
  label: PropTypes.string,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func,
};

export default Radio;
