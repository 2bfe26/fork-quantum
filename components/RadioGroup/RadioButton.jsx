import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import ButtonGroupLabel from '../shared/ButtonGroupLabel';
import HiddenInput from '../shared/HiddenInput';
import Icon from '../Icon';
import uniqId from '../shared/uniqId';

const ID_GENERATOR = uniqId('radio-button-');

const LockIcon = styled(Icon).attrs({
  name: 'lock',
})``;

const ButtonIcon = styled(Icon)`
  margin-right: 8px;
`;

const Wrapper = styled.div`
  ${({ inline }) =>
    inline &&
    `
    display: inline-block;
    margin-right: 16px;
    vertical-align: top;

    :last-child {
      margin-right: 0;
    }
  `}

  ${LockIcon} {
    font-size: 17px;
    margin-left: 8px;
  }
`;

class Radio extends React.Component {
  constructor(props) {
    super(props);

    const { id } = props;

    this._id = id || ID_GENERATOR.next().value;
  }

  render() {
    const {
      children,
      label,
      error,
      disabled,
      onChange,
      value,
      checked,
      icon,
      id,
      inline,
      ...rest
    } = this.props;
    const skin = checked ? 'primary' : 'secondary';

    return (
      <Wrapper inline={inline}>
        <HiddenInput
          type="radio"
          checked={checked}
          disabled={disabled}
          id={this._id}
          onChange={e => onChange({ value, label }, e)}
          value={value}
          skin={skin}
          error={error}
          {...rest}
        />
        <ButtonGroupLabel
          checked={checked}
          disabled={disabled}
          error={error}
          htmlFor={this._id}
          skin={skin}
        >
          {icon && <ButtonIcon name={icon} />}
          {children || label}
          {disabled && <LockIcon />}
        </ButtonGroupLabel>
      </Wrapper>
    );
  }
}

Radio.displayName = 'RadioGroup.Button';

Radio.defaultProps = {
  checked: false,
  children: undefined,
  disabled: false,
  error: false,
  icon: undefined,
  id: undefined,
  inline: false,
  label: undefined,
  onChange: () => {},
};

Radio.propTypes = {
  checked: PropTypes.bool,
  children: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  icon: PropTypes.string,
  id: PropTypes.string,
  inline: PropTypes.bool,
  label: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.string.isRequired,
};

export default Radio;
