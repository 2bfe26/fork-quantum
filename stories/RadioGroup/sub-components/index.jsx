import simpleRadioGroup from './simpleRadioGroup';
import propsRadioGroup from './propsRadioGroup';
import disabledRadio from './disabledRadio';
import buttonGroup from './buttonGroup';

export { simpleRadioGroup, propsRadioGroup, disabledRadio, buttonGroup };
