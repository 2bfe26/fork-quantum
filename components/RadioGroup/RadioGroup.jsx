import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Colors from '../Colors';
import { FieldGroup, Label } from '../shared';

import Radio from './Radio';

const Group = styled(FieldGroup).attrs({
  role: 'radiogroup',
})`
  display: flex;
  flex-direction: column;

  position: relative;
`;

const ErrorLabel = styled(Label)`
  color: ${Colors.ERROR['900']};
  cursor: text;
  font-weight: 600;
  font-style: italic;
  font-size: 14px;
  left: -3px;
  position: absolute;

  bottom: -30px;
`;

ErrorLabel.displayName = 'ErrorLabel';

class RadioGroup extends React.Component {
  static Radio = Radio;

  render() {
    const { children, name, error, options, ...rest } = this.props;

    const radioOptions = options.map(option =>
      Object.assign({}, option, { name }),
    );

    const listItems = children || radioOptions.map(Radio.create);

    return (
      <Group {...rest}>
        {listItems}
        {error && <ErrorLabel>{error}</ErrorLabel>}
      </Group>
    );
  }
}

/**
 * Group for Radio components.
 */
RadioGroup.defaultProps = {
  value: undefined,
  error: undefined,
  options: [],
  onChange: () => {},
};

RadioGroup.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
      value: PropTypes.string,
      disabled: PropTypes.bool,
    }),
  ),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  onChange: PropTypes.func,
  /** Initialize RadioGroup with a value */
  value: PropTypes.string,
  name: PropTypes.string.isRequired,
  error: PropTypes.string,
};

export default RadioGroup;
