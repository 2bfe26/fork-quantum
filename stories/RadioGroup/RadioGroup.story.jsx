import React from 'react';
import PropTypes from 'prop-types';
import { storiesOf } from '@storybook/react';
import {
  Heading,
  AutoPropsApi,
  HowToImport,
  CodeExample,
  TabbedView,
  Tab,
} from '@catho-private/quantum-storybook-ui';

import RadioGroup from '../../components/RadioGroup';
import { Col, Row } from '../../components/Grid';

import * as samples from './sub-components';

const RadioGroupExample = ({ code, component }) => (
  <>
    <Col small={5}>
      <CodeExample component={{}} code={code} showTitle={false} />
    </Col>
    <Col small={7}>{component}</Col>
  </>
);

RadioGroupExample.propTypes = {
  code: PropTypes.string.isRequired,
  component: PropTypes.node.isRequired,
};

const stories = storiesOf('3. Forms', module);
stories.add('Radio Button', () => (
  <Heading name="Radio" title="<RadioGroup />">
    <TabbedView>
      <Tab title="Usage">
        <HowToImport importModules="RadioGroup" componentName="RadioGroup" />

        <p>
          We provide two components to use Radio Buttons:{' '}
          <code>{'<RadioGroup />'}</code> and{' '}
          <code>{'<RadioGroup.Radio />'}</code>
        </p>

        <p>
          With a very simple API, you can set an radio group using just a array,
          such as <code>RadioGroup.Radio</code>:
        </p>
        <br />

        <Row>
          <RadioGroupExample {...samples.simpleRadioGroup} />
        </Row>
        <br />

        <p>
          Also, you can set some properties to <code>RadioGroup</code>, such as
          the selected <code>value</code>, <code>error</code> and{' '}
          <code>onChange</code> function
        </p>
        <br />

        <Row>
          <RadioGroupExample {...samples.propsRadioGroup} />
        </Row>
        <br />

        <p>
          For <code>RadioGroup.Radio</code>, and option shape as well, you can
          add the <code>disable</code> property too
        </p>

        <br />
        <Row>
          <RadioGroupExample {...samples.disabledRadio} />
        </Row>
      </Tab>
      <Tab title="API">
        <AutoPropsApi component={RadioGroup} title="RadioGroup" />
        <AutoPropsApi component={RadioGroup.Radio} title="RadioGroup.Radio" />
      </Tab>
    </TabbedView>
  </Heading>
));
