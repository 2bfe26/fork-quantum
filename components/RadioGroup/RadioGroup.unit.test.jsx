import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import diff from 'jest-diff';
import stripAnsi from 'strip-ansi';
import RadioGroup from './RadioGroup';

describe('<RadioGroup />', () => {
  describe('snapshot', () => {
    it('simple with options', () => {
      const options = [
        { value: 'Foo', label: 'Foo' },
        { value: 'Bar', label: 'Bar' },
        { value: 'Baz', label: 'Baz' },
      ];

      const component = <RadioGroup name="foo" options={options} />;
      expect(renderer.create(component)).toMatchSnapshot();
    });

    it('simple with <RadioGroup.Radio />', () => {
      const component = (
        <RadioGroup name="groceries">
          <RadioGroup.Radio value="Foo">Foo</RadioGroup.Radio>
          <RadioGroup.Radio value="Bar">Bar</RadioGroup.Radio>
          <RadioGroup.Radio value="Baz">Baz</RadioGroup.Radio>
        </RadioGroup>
      );
      expect(renderer.create(component)).toMatchSnapshot();
    });

    it('should render the same', () => {
      const options = [
        { value: 'Foo', label: 'Foo' },
        { value: 'Bar', label: 'Bar' },
        { value: 'Baz', label: 'Baz' },
      ];

      const snapOptions = renderer
        .create(<RadioGroup name="foo" options={options} />)
        .toJSON();

      const snapComposable = renderer
        .create(
          <RadioGroup name="foo">
            <RadioGroup.Radio value="Foo">Foo</RadioGroup.Radio>
            <RadioGroup.Radio value="Bar">Bar</RadioGroup.Radio>
            <RadioGroup.Radio value="Baz">Baz</RadioGroup.Radio>
          </RadioGroup>,
        )
        .toJSON();

      expect(
        `Snapshot Diff:\n ${stripAnsi(diff(snapOptions, snapComposable))}`,
      ).toMatchSnapshot();
    });
  });

  describe('<RadioGroup.Radio />', () => {
    it('simple case', () => {
      const component = (
        <RadioGroup name="foo">
          <RadioGroup.Radio value="Foo">Foo</RadioGroup.Radio>
          <RadioGroup.Radio value="Bar">Bar</RadioGroup.Radio>
          <RadioGroup.Radio value="Baz">Baz</RadioGroup.Radio>
        </RadioGroup>
      );
      const wrapper = shallow(component);
      const radios = wrapper.find(RadioGroup.Radio);

      expect(radios).toHaveLength(3);
    });

    it('every <RadioGroup.Radio /> should have the <RadioGroup /> name', () => {
      const component = (
        <RadioGroup name="foo">
          <RadioGroup.Radio value="Foo">Foo</RadioGroup.Radio>
          <RadioGroup.Radio value="Bar">Bar</RadioGroup.Radio>
          <RadioGroup.Radio value="Baz">Baz</RadioGroup.Radio>
        </RadioGroup>
      );
      const wrapper = mount(component);
      const radios = wrapper.find(RadioGroup.Radio);

      radios.map(radio =>
        expect(radio.prop('name')).toBe(wrapper.prop('name')),
      );
    });

    it('should have selected prop', () => {
      const component = (
        <RadioGroup name="foo" value="Bar">
          <RadioGroup.Radio value="Foo">Foo</RadioGroup.Radio>
          <RadioGroup.Radio value="Bar">Bar</RadioGroup.Radio>
          <RadioGroup.Radio value="Baz">Baz</RadioGroup.Radio>
        </RadioGroup>
      );
      const wrapper = shallow(component);
      const radios = wrapper.find(RadioGroup.Radio);

      const radioBar = radios.at(1);
      expect(radioBar.prop('checked')).toBe(true);
    });

    it('should call onChange on every <RadioGroup.Radio />', () => {
      const onChangeMock = jest.fn();

      const component = (
        <RadioGroup name="foo" onChange={onChangeMock}>
          <RadioGroup.Radio value="Foo">Foo</RadioGroup.Radio>
          <RadioGroup.Radio value="Bar">Bar</RadioGroup.Radio>
          <RadioGroup.Radio value="Baz">Baz</RadioGroup.Radio>
        </RadioGroup>
      );

      const wrapper = shallow(component);
      const radios = wrapper.find(RadioGroup.Radio);

      radios.map(radio => radio.simulate('change'));

      expect(onChangeMock).toBeCalledTimes(3);
    });
  });
});
