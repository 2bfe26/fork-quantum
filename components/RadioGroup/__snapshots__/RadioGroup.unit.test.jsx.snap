// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup /> Should match the snapshot 1`] = `
.c1 {
  cursor: pointer;
  display: inline-block;
}

.c1:before {
  content: ' ';
  border-width: 1.5px;
  border-style: solid;
  border-color: #999999;
  border-radius: 50%;
  display: inline-block;
  height: 16px;
  width: 16px;
  margin-right: 5px;
  position: relative;
  top: 4px;
}

.c1[aria-checked='true']:before {
  background-color: #1355d0;
  border-color: #1355d0;
  box-shadow: inset 0 0 0 3.5px #FFFFFF;
}

.c1[aria-checked='true']:hover:before,
.c1[aria-checked='true']:focus:before {
  box-shadow: inset 0 0 0 3.5px #FFFFFF,0 2px 6px 0 rgba(19,85,208,0.5);
}

.c1:hover:before,
.c1:focus:before {
  border-color: #1355d0;
  box-shadow: 0 2px 6px 0 rgba(19,85,208,0.5);
}

.c1[disabled] {
  color: #999999;
}

.c1[disabled]:before {
  background-color: #cccccc;
}

.c1[disabled]:hover {
  cursor: not-allowed;
}

.c1[disabled]:hover:before,
.c1[disabled]:focus:before {
  border-color: #999999;
  box-shadow: none;
}

.c1[aria-checked='true']:disabled:before {
  background-color: #999999;
  border-color: #999999;
  box-shadow: inset 0 0 0 3.5px #FFFFFF;
}

.c0 {
  position: relative;
  margin-bottom: 20px;
  min-width: 250px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
}

<div
  className="c0"
  onChange={[Function]}
  role="radiogroup"
>
  <div
    aria-checked={false}
    className="c1"
    disabled={false}
    name="group"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="radio"
    tabIndex={0}
    value="Foo"
  >
    Foo
  </div>
  <div
    aria-checked={false}
    className="c1"
    disabled={false}
    name="group"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="radio"
    tabIndex={-1}
    value="Bar"
  >
    Bar
  </div>
  <div
    aria-checked={false}
    className="c1"
    disabled={false}
    name="group"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="radio"
    tabIndex={-1}
    value="Baz"
  >
    Baz
  </div>
  
</div>
`;
